pkg/apis/core/types.go

```diff
@@ -3553,17 +3575,32 @@ type ServiceSpec struct {
 	// More info: https://kubernetes.io/docs/concepts/services-networking/service/
 	Selector map[string]string
 
-	// ClusterIP is the IP address of the service and is usually assigned
-	// randomly by the master. If an address is specified manually and is not in
-	// use by others, it will be allocated to the service; otherwise, creation
-	// of the service will fail. This field can not be changed through updates.
-	// Valid values are "None", empty string (""), or a valid IP address. "None"
-	// can be specified for headless services when proxying is not required.
-	// Only applies to types ClusterIP, NodePort, and LoadBalancer. Ignored if
-	// type is ExternalName.
-	// More info: https://kubernetes.io/docs/concepts/services-networking/service/#virtual-ips-and-service-proxies
+	// ClusterIPs identifies all the ClusterIPs assigned to or requested for  this
+	// service. ClusterIPs are assigned or reserved  based on the values of
+	// Service.Spec.IPFamilies. Maximum of two entries (dualstack IPs) are
+	// allowed in ClusterIPs. The IPFamily of each ClusterIP must match
+	// values provided in Service.Spec.IPFamilies
+	// +optional
+	ClusterIPs []string
+
+	// IPFamilies identifies all the IPFamilies assigned to or requested for
+	// this Service. If value was not provided for IPFamilies it will be defaulted
+	// based on the cluster configuration and the value of Service.Spec.IPFamilyPolicy.
+	// Maximum of two values (dualstack IPFamilies) are allowed in IPFamilies.
 	// +optional
-	ClusterIP string
+	IPFamilies []IPFamily
+
+	// IPFamilyPolicy represents the dualstack-ness requested or required by this
+	// Service. If there is no value provided, then this Service will be considered
+	// SingleStack (single IPFamily). Services can be SingleStack (single IPFamily)
+	// PreferedDualStack (two dualstack IPFamilies on dualstack clusters) or (single
+	// IPFamily on singlestack clusters). Services can also be RequireDualStack where
+	// it will have two dualstack IPFamilies on dualstack configured clusters otherwise
+	// (one a single stack cluster) it will fail. The IPFamilies and ClusterIPs assigned
+	// to this service can be controlled by Service.Spec.IPFamilies and Service.Spec.ClusterIPs
+	// respectively
+	// +optional
+	IPFamilyPolicy *IPFamilyPolicyType
```

Comment:
Do we want to talk about mutability?  IPFamilyPolicy is not mutable (or is it?) but IPFamilies in a very specific way.

Reply:
`IPFamilyPolicy` is mutable. We allow upgrades and downgrades to and from `DualStack`.

Reply:
I added the immutability comments on IPFamilies. IPFamilyPolicy has no immutability rules it can change from one value to another as needed.